chapter_12_dependency_injection.asciidoc

```diff
@@ -503,6 +530,29 @@ fine although it makes pylint unhappy.  You might also check out
 https://pypi.org/project/punq/[Punq], as written by Bob himself.
 
 // TODO: discuss/define what a DI container is
+//
+// (EJ2) based on the whatsapp thread, what I'd imagined is:
+// ===
+// class MessageBus:
```

Comment:
yes this is what most of the messagebuses at made look like.  i'll add it in a sidebar maybe?

Reply:
Sidebar snippet would be nice, and maybe it would give you the chance to talk about DI containers.  As a side note there is one more refinement I would consider doing here IRL:
```
# allocation/service_layer/__init__.py
def register_handlers(bus):
    bus.register_event_handler(FooEvent, handlers.handle_foo)
           .register_event_handler(...)

# allocation/bootstrap
def bootstrap(...):
    ...
     bus = MessageBus(uow)
    service_layer.register_handlers(bus)
```
Then you can package up a set of "reasonable default" handlers, and/oor alternative handler configurations that are selectable at run time.

Reply:
I wonder why you and @bobthemighty are so keen on having `bus.register_handler`?  I thought we wanted the bus to have _less_ responsibility for doing DI?

Reply:
DI and service location are different things. It's a somewhat subtle point.

Consider:

```

container = punq.container()
container.register(UnitOfWork, SqlAlchemyUnitOfWork)

class DoThing:

    def __init__(self, uow: UnitOfWork):
        ....


class MessageBus:

    def __init__(self, container):
        ....

    def register_handler(self, msg_type, handler_type):
        self.container.register(MessageHandler[msg_type], handler_type)

    def handle(self, msg):
        handlers = self.container.resolve(List[Handler[type(msg)]])
        for handler in handlers:
            handler(msg)


bus = MessageBus(container)
bus.register(ThingHappened, DoThing)
```

In this code, the bus acts as a message registry, but knows nothing about how to actually construct any particular handler. It can `locate` a service, but can't inject the dependencies of that service.

Reply:
In my mind, `register_handler` (1) hides the messy details of routing + dispatching events to their matching handlers. (2) It should do nothing but routing + dispatching, and be able to treat handlers uniformly (which corresponds to bob's point about not caring how to construct a handler).

Injecting the handler dictionaries accomplishes (2) but not (1), because the client has to be know about the internal details of message dispatch to correctly construct the dictionary values.

The `inject_dependencies` snippet style kind of does (1) but not (2).  IIRC, you still have to inject all the dependencies of all handlers into the constructor correctly, and won't find out if you've made a mistake until runtime.  Or you have to maintain a constructor signature that knows about all possible abstract dependencies of the handlers, which couples the bus to domain model details.  It also does not easily generalize to account for dependency chains (the handler dependencies' dependencies).  You could write code to account for that, but at that point I think you are better off using something like punq.

Reply:
oo, dependency chains! i hadn't thought of that.  i will add a mention.

I don't understand the point about the constructor signature (which constructor?) needing to know about all dependencies (handlers already know their dependencies don't they?  unless there's dependency chanis as you say...) or coupling the bus to the domain model (the domain model has no deps?).